File: 56-RADStudio10.3/AndroidAPILevelUpgrade_Demos/Object Pascal/Multi-Device Samples/Media/MusicPlayer/MusicPlayer.Android.pas
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

unit MusicPlayer.Android;

interface
{$IFDEF ANDROID}
uses
  FMX.Graphics,
  MusicPlayer.Utils,
  System.IoUtils, System.SysUtils, System.Classes,
  FMX.Types, FMX.Platform.Android,
  Androidapi.JNI.Os, Androidapi.JNI.Net,
  Androidapi.JNIBridge, Androidapi.JNI.JavaTypes, Androidapi.JNI.GraphicsContentViewText,
  Androidapi.JNI.Media, Androidapi.JNI.Provider, Androidapi.Helpers, Androidapi.JNI.App;
type
  TMusicPlayer = class
  private
  type
    TProcessThread = class (TThread)
    private
      [weak] FMusicPlayer: TMusicPlayer;
      FOnProcessPlay: TOnProcessPlayEvent;
      procedure DoProcessPlay;
    public
      constructor Create(CreateSuspended: Boolean; AMusicPlayer: TMusicPlayer; processHandler: TOnProcessPlayEvent);
      destructor Destroy; override;
      procedure Execute; override;
    end;
  protected
    class var FInstance: TMusicPlayer;
  private
    FCurrentIndex: Integer;
    FPlaylist: TArray<TMPSong>;
    FAlbums: TArray<TMPAlbum>;
    FMusicPlayer: JMediaPlayer;
    FPlayBackState: TMPPlaybackState;
    FRepeatMode: TMPRepeatMode;
    FShuffleMode: Boolean;
    FDefaultAlbumImage: TBitmap;
    FOnSongChange: TOnSongChangeEvent;
    FOnProcessPlay: TOnProcessPlayEvent;
    constructor Create(AType: TMPControllerType = TMPControllerType.App);
    procedure DoOnSongChange(newIndex: Integer);
    procedure DoOnProcessPlay(newPos: Single);
    procedure SetVolume(const Value: Single);
    procedure SetTime(const Value: Single);
    procedure SetRepeatMode(const Value: TMPRepeatMode);
    procedure SetShuffleMode(const Value: Boolean);
    function GetVolume: Single;
    function GetTime: Single;
    function GetRepeatMode: TMPRepeatMode;
    function GetDuration: Single;
    function GetPlaybackState: TMPPlaybackState;
    function GetShuffleMode: Boolean;
  public
    destructor Destroy; override;
    class procedure SetPlayerType(AType: TMPControllerType);
    class function DefaultPlayer: TMusicPlayer;
    property CurrentIndex: Integer read FCurrentIndex;
    property Volume: Single read GetVolume write SetVolume;
    property Time: Single read GetTime write SetTime;
    property Duration: Single read GetDuration;
    property PlaybackState: TMPPlaybackState read GetPlaybackState;
    property ShuffleMode: Boolean read GetShuffleMode write SetShuffleMode;
    property RepeatMode: TMPRepeatMode read GetRepeatMode write SetRepeatMode;
    property Playlist: TArray<TMPSong> read FPlaylist;
    property Albums: TArray<TMPAlbum> read FAlbums;
    property DefaultAlbumImage: TBitmap read FDefaultAlbumImage write FDefaultAlbumImage;
    property OnSongChange: TOnSongChangeEvent read FOnSongChange write FOnSongChange;
    property OnProcessPlay: TOnProcessPlayEvent read FOnProcessPlay write FOnProcessPlay;
    function GetAlbums: TArray<string>;
    function GetSongs: TArray<string>;
    function GetSongsInAlbum(AName: string): TArray<string>;
    function IsPlaying: Boolean;
    function CanSkipBack: Boolean;
    function CanSkipForward: Boolean;
    procedure PlaySong(path: string);
    procedure PlayByIndex(Index: Integer);
    procedure Play;
    procedure Stop;
    procedure Pause;
    procedure Next;
    procedure Previous;
  end;
var
  NoArtBitmap: TBitmap;
{$ENDIF}
implementation
{$IFDEF ANDROID}
{ TMusicPlayer }

function TMusicPlayer.CanSkipBack: Boolean;
begin
  Result := (Length(FPlaylist) > 0) and (FCurrentIndex > 0) and
    (FPlayBackState in [TMPPlaybackState.Playing, TMPPlaybackState.Paused]);
end;

function TMusicPlayer.CanSkipForward: Boolean;
begin
  Result := False;
  if (Length(FPlaylist) = 0) or not (FPlayBackState in [TMPPlaybackState.Playing, TMPPlaybackState.Paused]) then
    Exit(Result);

  case RepeatMode of
    TMPRepeatMode.One:
      Result := FCurrentIndex in [Low(FPlaylist) .. High(FPlaylist)] ;
    TMPRepeatMode.Default,
    TMPRepeatMode.None:
      Result := FCurrentIndex in [Low(FPlaylist) .. High(FPlaylist)-1] ;
    TMPRepeatMode.All:
      Result := True;
  end;
end;

constructor TMusicPlayer.Create(AType: TMPControllerType);
begin
  MainActivity.setVolumeControlStream(TJAudioManager.JavaClass.STREAM_MUSIC);
  FMusicPlayer := TJMediaPlayer.Create;
  FPlayBackState := TMPPlaybackState.Stopped;
  FRepeatMode := TMPRepeatMode.All;
  FShuffleMode := False;
  FDefaultAlbumImage := TBitmap.CreateFromFile(TPath.Combine(TPath.GetDocumentsPath,'MusicNote.png'));
  TProcessThread.Create(True,self,DoOnProcessPlay).Start;
end;

class function TMusicPlayer.DefaultPlayer: TMusicPlayer;
begin
  if not Assigned(FInstance) then
    FInstance := TMusicPlayer.Create;
  Result := FInstance;
end;

destructor TMusicPlayer.Destroy;
begin
  FMusicPlayer.release;
end;

procedure TMusicPlayer.DoOnSongChange(newIndex: Integer);
begin
  if Assigned(FOnSongChange) then
    TThread.Queue(TThread.CurrentThread, procedure
      begin
        FOnSongChange(newIndex);
      end);
end;

procedure TMusicPlayer.DoOnProcessPlay(newPos: Single);
begin
  if Assigned(FOnProcessPlay) then
    TThread.Queue(TThread.CurrentThread, procedure
      begin
        FOnProcessPlay(newPos);
      end);
end;

function TMusicPlayer.GetAlbums: TArray<string>;
var
  projection: TJavaObjectArray<JString>;
  cursor: JCursor;
  artPath: String;
  tmpPath: String;
begin
  projection := TJavaObjectArray<JString>.Create(4);
  projection.Items[0] := TJAudio_AlbumColumns.JavaClass.ALBUM;
  projection.Items[1] := TJAudio_AlbumColumns.JavaClass.ARTIST;
  projection.Items[2] := StringToJString('_id');
  projection.Items[3] := TJAudio_AlbumColumns.JavaClass.ALBUM_ART;

  cursor := MainActivity.getContentResolver.query(
    TJAudio_Albums.JavaClass.EXTERNAL_CONTENT_URI,
    projection,
    nil,
    nil,
    nil);

  SetLength(Result, cursor.getCount);
  SetLength(FAlbums, cursor.getCount + 1);
  FAlbums[cursor.getCount] := TMPAlbum.AllMusicAlbum;
  while (cursor.moveToNext) do
  begin
    FAlbums[cursor.getPosition].Name := JStringToString(cursor.getString(0));
    FAlbums[cursor.getPosition].Artist := JStringToString(cursor.getString(1));
    FAlbums[cursor.getPosition].Album_ID := cursor.getInt(2);

    artPath := JStringToString(cursor.getString(3));
    if TFile.Exists(artPath) then
    begin
      try
        //Workaround for loading problems: copy to a file with correct extension.
        tmpPath := System.IOUtils.TPath.Combine(System.IOUtils.TPath.GetDocumentsPath,  'tmp.jpg');
        TFile.Copy(artPath, tmpPath);
        FAlbums[cursor.getPosition].Artwork := TBitmap.CreateFromFile(tmpPath);
        TFile.Delete(tmpPath);
      except
      end;
    end
    else
      FAlbums[cursor.getPosition].Artwork := FDefaultAlbumImage;

    Result[cursor.getPosition] := FAlbums[cursor.getPosition].Name;
  end;
  cursor.close;
end;

function TMusicPlayer.GetDuration: Single;
begin
  Result := FMusicPlayer.getDuration;
end;

function TMusicPlayer.GetPlaybackState: TMPPlaybackState;
begin
  Result := FPlayBackState;
end;

function TMusicPlayer.GetRepeatMode: TMPRepeatMode;
begin
  Result := FRepeatMode;
end;

function TMusicPlayer.GetShuffleMode: Boolean;
begin
  Result := FShuffleMode;
end;

function TMusicPlayer.GetSongs: TArray<string>;
var
  projection: TJavaObjectArray<JString>;
  cursor: JCursor;
  selection: JString;
begin
  selection := StringToJString(JStringToString(TJAudio_AudioColumns.JavaClass.IS_MUSIC) + ' != 0');

  projection := TJavaObjectArray<JString>.Create(5);

  projection.Items[0] := TJAudio_AudioColumns.JavaClass.ARTIST;
  projection.Items[1] := StringToJString('title');
  projection.Items[2] := StringToJString('_data');
  projection.Items[3] := TJAudio_AudioColumns.JavaClass.ALBUM;
  projection.Items[4] := TJAudio_AudioColumns.JavaClass.DURATION;

  cursor := MainActivity.getContentResolver.query(
    TJAudio_Media.JavaClass.EXTERNAL_CONTENT_URI,
    projection,
    selection,
    nil,
    nil);

  SetLength(Result,cursor.getCount);
  SetLength(FPlaylist, cursor.getCount);
  while (cursor.moveToNext) do
  begin
    FPlaylist[cursor.getPosition] := TMPSong.FromCursor(cursor);
    Result[cursor.getPosition] := Format('[%s]-[%s]', [FPlaylist[cursor.getPosition].Artist, FPlaylist[cursor.getPosition].Title]);
  end;
  cursor.close;
end;

function TMusicPlayer.GetSongsInAlbum(AName: string): TArray<string>;
var
  projection: TJavaObjectArray<JString>;
  cursor: JCursor;
  selection: JString;
begin
  if AName = TMPAlbum.AllMusicAlbum.Name then
  begin
    Result := GetSongs;
    Exit;
  end;

  selection := StringToJString(JStringToString(TJAudio_AudioColumns.JavaClass.IS_MUSIC) + ' != 0 and ' +
      JStringToString(TJAudio_AudioColumns.JavaClass.ALBUM) + ' = "' + AName + '"');

  projection := TJavaObjectArray<JString>.Create(5);

  projection.Items[0] := TJAudio_AudioColumns.JavaClass.ARTIST;
  projection.Items[1] := StringToJString('title');
  projection.Items[2] := StringToJString('_data');
  projection.Items[3] := TJAudio_AudioColumns.JavaClass.ALBUM;
  projection.Items[4] := TJAudio_AudioColumns.JavaClass.DURATION;

  cursor := MainActivity.getContentResolver.query(
    TJAudio_Media.JavaClass.EXTERNAL_CONTENT_URI,
    projection,
    selection,
    nil,
    nil);

  SetLength(Result,cursor.getCount);
  SetLength(FPlaylist, cursor.getCount);
  while (cursor.moveToNext) do
  begin
    FPlaylist[cursor.getPosition] := TMPSong.FromCursor(cursor);
    Result[cursor.getPosition] := Format('[%s]-[%s]', [FPlaylist[cursor.getPosition].Artist, FPlaylist[cursor.getPosition].Title]);
  end;
  cursor.close;
end;

function TMusicPlayer.GetTime: Single;
begin
  Result := FMusicPlayer.getCurrentPosition;
end;

function TMusicPlayer.GetVolume: Single;
var
  AudioManager: JAudioManager;
begin
  AudioManager := TJAudioManager.Wrap(MainActivity.getSystemService(TJContext.JavaClass.AUDIO_SERVICE));
  Result := AudioManager.getStreamVolume(TJAudioManager.JavaClass.STREAM_MUSIC);
  Result := Result / AudioManager.getStreamMaxVolume(TJAudioManager.JavaClass.STREAM_MUSIC);
end;

procedure TMusicPlayer.Next;
begin
  case RepeatMode of
    TMPRepeatMode.One:
    begin
      Time := 0;
      Play;
    end;
    TMPRepeatMode.Default,
    TMPRepeatMode.None:
      if CurrentIndex = Length(FPlaylist) - 1  then
        FMusicPlayer.Stop
      else
      begin
        if ShuffleMode then
          PlayByIndex(Random(Length(FPlaylist)))
        else
          PlayByIndex(FCurrentIndex + 1);
      end;
    TMPRepeatMode.All:
      if FCurrentIndex = Length(FPlaylist) - 1 then
        PlayByIndex(0)
      else
      begin
        if FShuffleMode then
          PlayByIndex(Random(Length(FPlaylist)))
        else
          PlayByIndex(FCurrentIndex + 1);
      end;
  end;
  DoOnSongChange(FCurrentIndex);
end;

procedure TMusicPlayer.Pause;
begin
  FMusicPlayer.pause;
  FPlayBackState := TMPPlaybackState.Paused;
end;

procedure TMusicPlayer.Play;
begin
  if FPlayBackState = TMPPlaybackState.Stopped  then
    FMusicPlayer.prepare;
  FMusicPlayer.start;
  FPlayBackState := TMPPlaybackState.Playing;
end;

procedure TMusicPlayer.PlayByIndex(Index: Integer);
begin
  FCurrentIndex := Index;
  PlaySong(FPlaylist[FCurrentIndex].Path);
end;

function TMusicPlayer.IsPlaying: Boolean;
begin
  Result := FMusicPlayer.isPlaying;
end;

procedure TMusicPlayer.PlaySong(path: string);
begin
  Stop;
  FMusicPlayer.reset;
  FMusicPlayer.setDataSource(StringToJString(path));
  Play;
end;

procedure TMusicPlayer.Previous;
begin
  if (FCurrentIndex > 0) and (FCurrentIndex < Length(FPlaylist)) then
  begin
    PlayByIndex(FCurrentIndex -1);
    DoOnSongChange(FCurrentIndex);
  end;
end;

class procedure TMusicPlayer.SetPlayerType(AType: TMPControllerType);
begin
  // Do nothing
end;

procedure TMusicPlayer.SetRepeatMode(const Value: TMPRepeatMode);
begin
  FRepeatMode := Value;
end;

procedure TMusicPlayer.SetShuffleMode(const Value: Boolean);
begin
  FShuffleMode := Value;
end;

procedure TMusicPlayer.SetTime(const Value: Single);
begin
  FMusicPlayer.seekTo(Trunc(Value));
end;

procedure TMusicPlayer.SetVolume(const Value: Single);
var
  AudioManager: JAudioManager;
begin
  AudioManager := TJAudioManager.Wrap(MainActivity.getSystemService(TJContext.JavaClass.AUDIO_SERVICE));
  AudioManager.setStreamVolume(TJAudioManager.JavaClass.STREAM_MUSIC,
    Round(AudioManager.getStreamMaxVolume(TJAudioManager.JavaClass.STREAM_MUSIC) * Value), 0);
end;

procedure TMusicPlayer.Stop;
begin
  if FPlayBackState in [TMPPlaybackState.Playing, TMPPlaybackState.Paused] then
    FMusicPlayer.seekTo(0);
  FPlayBackState := TMPPlaybackState.Stopped;
  FMusicPlayer.stop;
  while FMusicPlayer.isPlaying do
    sleep(10);
  DoOnProcessPlay(0);
end;

{ TMusicPlayer.TProcessThread }

constructor TMusicPlayer.TProcessThread.Create(CreateSuspended: Boolean;
  AMusicPlayer: TMusicPlayer; processHandler: TOnProcessPlayEvent);
begin
  inherited Create(CreateSuspended);
  FOnProcessPlay := processHandler;
  FMusicPlayer := AMusicPlayer;
end;

destructor TMusicPlayer.TProcessThread.Destroy;
begin
  FMusicPlayer := nil;
  inherited;
end;

procedure TMusicPlayer.TProcessThread.Execute;
begin
  inherited;
  while Assigned(FMusicPlayer) do
  begin
    case FMusicPlayer.PlaybackState of
      Playing: DoProcessPlay;
      Stopped,
      Paused,
      Interrupted,
      SeekingForward,
      SeekingBackward: sleep(200);
    end;
  end;
end;

procedure TMusicPlayer.TProcessThread.DoProcessPlay;
var
  currentPos: Single;
begin
  currentPos := FMusicPlayer.Time;
  if Assigned(FOnProcessPlay) then
    FOnProcessPlay((currentPos/FMusicPlayer.Duration) * 100);


  if FMusicPlayer.IsPlaying then
    Sleep(200)
  else
    FMusicPlayer.Next;


end;
{$ENDIF}
end.

File: 56-RADStudio10.3/AndroidAPILevelUpgrade_Demos/CPP/Multi-Device Samples/Media/MusicPlayer/MusicPlayerAndroid.cpp
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

#pragma hdrstop

#include "MusicPlayerAndroid.h"
// ---------------------------------------------------------------------------
#ifdef __ANDROID__
using namespace Musicplayer::Utils;

namespace Musicplayer {
namespace Android {
Fmx::Graphics::TBitmap * NoArtBitmap;
TMusicPlayer* TMusicPlayer::FInstance;

inline __fastcall TMusicPlayer::TMusicPlayer(Utils::TMPControllerType AType) {
	MainActivity()->setVolumeControlStream(TJAudioManager::JavaClass->STREAM_MUSIC);
	FMusicPlayer = TJMediaPlayer::Create();
	FPlayBackState = Utils::TMPPlaybackState::Stopped;
	FRepeatMode = Utils::TMPRepeatMode::All;
	FShuffleMode = false;
	FDefaultAlbumImage =
		new TBitmap(TPath::Combine(TPath::GetDocumentsPath(),
		"MusicNote.png"));
	(new TProcessThread(true, this, DoOnProcessPlay))->Start();
}

__fastcall TMusicPlayer::~TMusicPlayer(void) {
	FMusicPlayer->release();
}

void __fastcall TMusicPlayer::DoOnSongChange(int newIndex) {
	if (FOnSongChange) {
		TThread::Queue(TThread::CurrentThread,
			new MyThreadMethod<TOnSongChangeEvent, int>(FOnSongChange,
			newIndex));
	}
}

void __fastcall TMusicPlayer::DoOnProcessPlay(float newPos) {
	if (FOnProcessPlay) {
		TThread::Queue(TThread::CurrentThread,
			new MyThreadMethod<TOnProcessPlayEvent, float>
			(FOnProcessPlay, newPos));
	}
}

void __fastcall TMusicPlayer::SetVolume(const float Value) {
	_di_JAudioManager AudioManager;
	_di_JObject obj = MainActivity()->getSystemService(TJContext::JavaClass->AUDIO_SERVICE);
	AudioManager = TJAudioManager::Wrap((_di_IJavaInstance)obj);

	AudioManager->setStreamVolume(TJAudioManager::JavaClass->STREAM_MUSIC,
		(int)(AudioManager->getStreamMaxVolume(TJAudioManager::JavaClass->STREAM_MUSIC) * Value), 0);
}

void __fastcall TMusicPlayer::SetTime(const float Value) {
	FMusicPlayer->seekTo((int) Value);
}

void __fastcall TMusicPlayer::SetRepeatMode
	(const Utils::TMPRepeatMode Value) {
	FRepeatMode = Value;
}

void __fastcall TMusicPlayer::SetShuffleMode(const bool Value) {
	FShuffleMode = Value;
}

float __fastcall TMusicPlayer::GetVolume(void) {
	_di_JAudioManager AudioManager;
	_di_JObject obj = MainActivity()->getSystemService(TJContext::JavaClass->AUDIO_SERVICE);
	AudioManager = TJAudioManager::Wrap((_di_IJavaInstance)obj);
	int result = AudioManager->getStreamVolume(TJAudioManager::JavaClass->STREAM_MUSIC);
	return (result * 1.0) / AudioManager->getStreamMaxVolume(TJAudioManager::JavaClass->STREAM_MUSIC);
}

float __fastcall TMusicPlayer::GetTime(void) {
	return FMusicPlayer->getCurrentPosition();
}

Utils::TMPRepeatMode __fastcall TMusicPlayer::GetRepeatMode(void) {
	return FRepeatMode;
}

float __fastcall TMusicPlayer::GetDuration(void) {
	return FMusicPlayer->getDuration();
}

Utils::TMPPlaybackState __fastcall TMusicPlayer::GetPlaybackState(void)
{
	return FPlayBackState;
}

bool __fastcall TMusicPlayer::GetShuffleMode(void) {
	return FShuffleMode;
}

void __fastcall TMusicPlayer::SetPlayerType
	(Utils::TMPControllerType AType) {
	// Do nothing
}

TMusicPlayer* __fastcall TMusicPlayer::DefaultPlayer() {
	if (!FInstance) {
		FInstance = new TMusicPlayer(TMPControllerType::App);
	}
	return FInstance;
}

System::DynamicArray<System::UnicodeString> __fastcall
TMusicPlayer::GetAlbums(void) {
	TJavaObjectArray__1<_di_JString> *projection;
	_di_JCursor cursor;
	System::DynamicArray<System::UnicodeString> result;
    String artPath, tmpPath;

	projection = new TJavaObjectArray__1<_di_JString>(4);
	projection->Items[0] = TJAudio_AlbumColumns::JavaClass->ALBUM;
	projection->Items[1] = TJAudio_AlbumColumns::JavaClass->ARTIST;
	projection->Items[2] = StringToJString("_id");
	projection->Items[3] = TJAudio_AlbumColumns::JavaClass->ALBUM_ART;

	cursor = Fmx::Platform::Android::MainActivity()->getContentResolver
		()->query(TJAudio_Albums::JavaClass->EXTERNAL_CONTENT_URI,
		projection, NULL, NULL, NULL);

	result.set_length(cursor->getCount());
	FAlbums.set_length(cursor->getCount() + 1);
	FAlbums[cursor->getCount()] = Utils::TMPAlbum::AllMusicAlbum();
	while (cursor->moveToNext()) {
		Log::d("Albums: CursorPosition :" +
			IntToStr(cursor->getPosition()) + " from " +
			IntToStr(cursor->getCount()));
		FAlbums[cursor->getPosition()].Name =
			JStringToString(cursor->getString(0));
		FAlbums[cursor->getPosition()].Artist =
			JStringToString(cursor->getString(1));
		FAlbums[cursor->getPosition()].Album_ID = cursor->getInt(2);

		artPath = JStringToString(cursor->getString(3));

		if( TFile::Exists(artPath)) {
			try {
				//Workaround for loading problems: copy to a file with correct extension.
				tmpPath = System::Ioutils::TPath::Combine(System::Ioutils::TPath::GetDocumentsPath(), L"tmp.jpg");
				TFile::Copy(artPath, tmpPath);
				FAlbums[cursor->getPosition()].Artwork = new TBitmap(tmpPath);
				TFile::Delete(tmpPath);
			}
			catch(...) {
            }
		}
		else {
			FAlbums[cursor->getPosition()].Artwork = FDefaultAlbumImage;
		}
	}
	cursor->close();

	return result;
}

System::DynamicArray<System::UnicodeString>__fastcall
	TMusicPlayer::GetSongs(void) {
	TJavaObjectArray__1<_di_JString> *projection;
	_di_JCursor cursor;
	_di_JString selection;
	System::DynamicArray<System::UnicodeString>result;

	selection =
		StringToJString(JStringToString
		(TJAudio_AudioColumns::JavaClass->IS_MUSIC) + " != 0");

	projection = new TJavaObjectArray__1<_di_JString>(5);
	projection->Items[0] = TJAudio_AudioColumns::JavaClass->ARTIST;
	projection->Items[1] = StringToJString("title");
	projection->Items[2] = StringToJString("_data");
	projection->Items[3] = TJAudio_AudioColumns::JavaClass->ALBUM;
	projection->Items[4] = TJAudio_AudioColumns::JavaClass->DURATION;

	cursor = Fmx::Platform::Android::MainActivity()->getContentResolver
		()->query(TJAudio_Media::JavaClass->EXTERNAL_CONTENT_URI,
		projection, selection, NULL, NULL);

	result.set_length(cursor->getCount());
	FPlaylist.set_length(cursor->getCount());

	while (cursor->moveToNext()) {
		FPlaylist[cursor->getPosition()] =
			Utils::TMPSong::FromCursor(cursor);
		result[cursor->getPosition()] =
			"[" + FPlaylist[cursor->getPosition()].Artist + "]-[" +
			FPlaylist[cursor->getPosition()].Title;
	}
	cursor->close();
	return result;
}

System::DynamicArray<System::UnicodeString>__fastcall
	TMusicPlayer::GetSongsInAlbum(System::UnicodeString AName) {
	if (AName == TMPAlbum::AllMusicAlbum().Name) {
		return GetSongs();
	}

	TJavaObjectArray__1<_di_JString> *projection;
	_di_JCursor cursor;
	_di_JString selection;
	System::DynamicArray<System::UnicodeString>result;

	selection =
		StringToJString(JStringToString
		(TJAudio_AudioColumns::JavaClass->IS_MUSIC) + " != 0 and " +
		JStringToString(TJAudio_AudioColumns::JavaClass->ALBUM) +
		" = \"" + AName + "\"");

	projection = new TJavaObjectArray__1<_di_JString>(5);
	projection->Items[0] = TJAudio_AudioColumns::JavaClass->ARTIST;
	projection->Items[1] = StringToJString("title");
	projection->Items[2] = StringToJString("_data");
	projection->Items[3] = TJAudio_AudioColumns::JavaClass->ALBUM;
	projection->Items[4] = TJAudio_AudioColumns::JavaClass->DURATION;

	cursor = Fmx::Platform::Android::MainActivity()->getContentResolver
		()->query(TJAudio_Media::JavaClass->EXTERNAL_CONTENT_URI,
		projection, selection, NULL, NULL);

	result.set_length(cursor->getCount());
	FPlaylist.set_length(cursor->getCount());

	while (cursor->moveToNext()) {
		FPlaylist[cursor->getPosition()] =
			Utils::TMPSong::FromCursor(cursor);
		result[cursor->getPosition()] =
			"[" + FPlaylist[cursor->getPosition()].Artist + "]-[" +
			FPlaylist[cursor->getPosition()].Title;
	}
	cursor->close();
	return result;
}

bool TMusicPlayer::IsPlaying() {
	return FMusicPlayer->isPlaying();
}

bool __fastcall TMusicPlayer::CanSkipBack()
{
	switch (FRepeatMode) {
		case Utils::TMPRepeatMode::Default:
		case Utils::TMPRepeatMode::None: return FCurrentIndex > FPlaylist.Low;
	default:
		return true;
	}
}

bool __fastcall TMusicPlayer::CanSkipForward()
{
   switch (FRepeatMode) {
		case Utils::TMPRepeatMode::Default:
		case Utils::TMPRepeatMode::None: return FCurrentIndex < FPlaylist.High;
	default:
		return true;
	}
}

void __fastcall TMusicPlayer::PlaySong(System::UnicodeString path) {
	Stop();
	FMusicPlayer->reset();
	FMusicPlayer->setDataSource(StringToJString(path));
	Play();
}

void __fastcall TMusicPlayer::PlayByIndex(unsigned Index) {
	FCurrentIndex = Index;
	PlaySong(FPlaylist[FCurrentIndex].Path);
}

void __fastcall TMusicPlayer::Play(void) {
	if (FPlayBackState == Utils::TMPPlaybackState::Stopped)
		FMusicPlayer->prepare();
	FMusicPlayer->start();
	FPlayBackState = Utils::TMPPlaybackState::Playing;
}

void __fastcall TMusicPlayer::Stop(void) {
	if (FPlayBackState == Utils::TMPPlaybackState::Playing ||
		FPlayBackState == Utils::TMPPlaybackState::Paused) {
		FMusicPlayer->seekTo(0);
	}
	FPlayBackState = Utils::TMPPlaybackState::Stopped;
	FMusicPlayer->stop();
	while (FMusicPlayer->isPlaying()) {
		Sleep(10);
	}
	DoOnProcessPlay(0);
}

void __fastcall TMusicPlayer::Pause(void) {
	FMusicPlayer->pause();
	FPlayBackState = Utils::TMPPlaybackState::Paused;
}

void __fastcall TMusicPlayer::Next(void) {
	switch (FRepeatMode) {
	case Utils::TMPRepeatMode::One:
		Time = 0;
		Play();
		break;
	case Utils::TMPRepeatMode::None:
	case Utils::TMPRepeatMode::Default:
		if (FCurrentIndex == FPlaylist.Length - 1) {
			FMusicPlayer->stop();
		}
		else {
			if (FShuffleMode)
				PlayByIndex(Random(FPlaylist.Length));
			else
				PlayByIndex(FCurrentIndex + 1);
		}
	case Utils::TMPRepeatMode::All:
		if (FCurrentIndex == FPlaylist.Length - 1) {
			PlayByIndex(0);
		}
		else {
			if (FShuffleMode)
				PlayByIndex(Random(FPlaylist.Length));
			else
				PlayByIndex(FCurrentIndex + 1);
		}
	default:
		break;
	}
	DoOnSongChange(FCurrentIndex);
}

void __fastcall TMusicPlayer::Previous(void) {
	if (FCurrentIndex > 0 && FCurrentIndex < FPlaylist.Length) {
		PlayByIndex(FCurrentIndex - 1);
		DoOnSongChange(FCurrentIndex);
	}
}

// ------------ TMusicPlayer::TProcessThread --------------------------------
__fastcall TMusicPlayer::TProcessThread::TProcessThread
	(bool CreateSuspended, TMusicPlayer* AMusicPlayer,
	Musicplayer::Utils::TOnProcessPlayEvent processHandler)
	: TThread(CreateSuspended) {
	FMusicPlayer = AMusicPlayer;
	FOnProcessPlay = processHandler;
}

// ---------------------------------------------------------------------------
__fastcall TMusicPlayer::TProcessThread::~TProcessThread(void) {
	FMusicPlayer = nullptr;
}

// ---------------------------------------------------------------------------
void __fastcall TMusicPlayer::TProcessThread::Execute(void) {
	while (FMusicPlayer) {
		Log::d("Process Thread: Execute");
		switch (FMusicPlayer->PlaybackState) {
		case TMPPlaybackState::Playing:
			ProcessPlay();
			break;
		case TMPPlaybackState::Stopped:
		case TMPPlaybackState::Paused:
		case TMPPlaybackState::Interrupted:
		case TMPPlaybackState::SeekingForward:
		case TMPPlaybackState::SeekingBackward:
			Sleep(200);
			break;
		}
	}
}

// ---------------------------------------------------------------------------
void __fastcall TMusicPlayer::TProcessThread::ProcessPlay(void) {
	float currentPos = FMusicPlayer->Time;
	if (FOnProcessPlay) {
		Log::d("Process Thread: ProcessPlay: call FOnProcessPlay");
		FOnProcessPlay((currentPos / FMusicPlayer->Duration) * 100);
	}

	if (FMusicPlayer->IsPlaying()) {
		Sleep(200);
	}
	else {
		FMusicPlayer->Next();
	}
}
// ---------------------------------------------------------------------------

} /* namespace Android */
} /* namespace Musicplayer */
#endif
#pragma package(smart_init)
